### PR TITLE
feat: Introduce schema definition.

### PR DESCRIPTION
**Pull request by liurenjie1024**
Initial schema definition. `SchemaVisitor`, name indexes will come later.

### CODE REVIEW

File: crates/iceberg/src/error.rs [hunk 1]
```diff
@@ -326,7 +326,7 @@ Context:
 Source: networking error
 
 Backtrace:
-   0: iceberg::error::tests::generate_error_with_backtrace_enabled
+   0: std::backtrace_rs::backtrace::libunwind::trace
```
Comment (Xuanwo) [inline crates/iceberg/src/error.rs:329]: Seems this line is different on different platforms, we can only check the content to `0: `.

File: crates/iceberg/src/spec/datatypes.rs [hunk 2]
```diff
@@ -305,6 +325,52 @@ pub struct StructField {
...
+    }
+
+    /// Set the field's initial default value.
+    pub fn with_initial_default(mut self, value: impl ToString) -> Self {
+        self.initial_default = Some(value.to_string());
+        self
+    }
+
+    /// Set the field's initial default value.
+    pub fn with_write_default(mut self, value: impl ToString) -> Self {
```
Comment (Xuanwo) [inline crates/iceberg/src/spec/datatypes.rs:368]: write_default and initial_default seems should be a `value`?

**Review by Xuanwo — APPROVED** [review]
Mostly LGTM! Although, I feel that the usage of `once_cell` here is a bit over-designed. Is it possible that we may need to add new fields into `StructType`, for example, in schema evaluation?

**Comment by Xuanwo** [comment]
> The Python and java implementation treats schema as immutable data structures, I think we should also follow that. Mutable makes things complicated, especially when we have indexes. For example the name indexing following.

Got it, makes sense now.

**Comment by JanKaul** [comment]
Should we also include [identifier-field-ids](https://iceberg.apache.org/spec/#schemas)?

**Comment by JanKaul** [comment]
I don't entirely understand. Could you elaborate why we would need the visitor pattern for the identifier-field-ids? I was thinking about something similar to the [serialized representation](https://iceberg.apache.org/spec/#schemas).

File: crates/iceberg/src/spec/datatypes.rs [hunk 3]
```diff
@@ -252,15 +253,34 @@ impl<'de> Deserialize<'de> for StructType {
...
-    /// Get structfield with certain id
+    /// Get struct field with certain id
     pub fn field_by_id(&self, id: i32) -> Option<&StructField> {
-        self.fields.get(*self.id_lookup.get(&id)?)
+        self.field_id_to_index(id).map(|idx| &self.fields[idx])
+    }
+
+    fn field_id_to_index(&self, field_id: i32) -> Option<usize> {
+        self.id_lookup
+            .get_or_init(|| {
```
Comment (Fokko) [inline crates/iceberg/src/spec/datatypes.rs:268]: Nice, this is very elegant

File: crates/iceberg/src/spec/schema.rs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+impl SchemaBuilder {
+    /// Add fields to schem builder
+    pub fn with_fields(mut self, fields: impl IntoIterator<Item = StructField>) -> Self {
+        self.fields.extend(fields.into_iter());
+        self
+    }
+
+    /// Set schema id.
+    pub fn with_schema_id(mut self, schema_id: i32) -> Self {
```
Comment (Fokko) [inline crates/iceberg/src/spec/schema.rs:46]: Schema ID can also be optional, then we would just set it to `0`.

File: crates/iceberg/src/spec/schema.rs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+const DEFAULT_SCHEMA_ID: i32 = 0;
+
+/// Defines schema in iceberg.
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Schema {
+    r#struct: StructType,
+    schema_id: i32,
+    highest_field_id: i32,
+}
```
Comment (Fokko) [inline crates/iceberg/src/spec/schema.rs:30]: I think the `identifier-field-ids` are missing: https://iceberg.apache.org/spec/#identifier-field-ids

**Review by Fokko — COMMENTED** [review]
Some small comments, but looks good 👍🏻

**Review by Fokko — APPROVED** [review]

**Pull request merged**